ledger/sandbox-on-x/src/classic/scala/com/daml/platform/ledger/SandboxServer.scala

```diff
@@ -104,25 +117,65 @@ final class SandboxServer(
-  ): AbstractResourceOwner[ResourceContext, List[Unit]] =
-    ResourceOwner.forFuture(() =>
-      Future.sequence(
-        config.damlPackages.map { file =>
-          val submissionId = Ref.SubmissionId.assertFromString(UUID.randomUUID().toString)
-          for {
-            dar <- Future.fromTry(DarParser.readArchiveFromFile(file).toTry)
-            _ <- writeService
-              .uploadPackages(submissionId, dar.all, None)(
-                LoggingContext.ForTesting,
-                NoOpTelemetryContext,
+  private def loadPackages(writeService: WriteService, indexService: IndexService)(implicit
+      executionContext: ExecutionContext,
+      system: ActorSystem,
+      loggingContext: LoggingContext,
+  ): AbstractResourceOwner[ResourceContext, Unit] =
+    ResourceOwner.forFuture(() => {
+      val packageUploadTimeout = 10.seconds
+
+      val packageSubmissionsTrackerMap = config.damlPackages.map { file =>
+        val uploadCompletionPromise = Promise[Unit]()
+        system.scheduler.scheduleOnce(packageUploadTimeout) {
+          // Package upload timeout, meant to allow fail-fast for testing
+          // TODO Remove once in-memory backend (e.g. H2) is deemed stable
+          uploadCompletionPromise.tryFailure(
+            new RuntimeException(s"Package upload timeout after $packageUploadTimeout")
+          )
+          ()
+        }
+        UUID.randomUUID().toString -> (file, uploadCompletionPromise)
+      }.toMap
+
+      val (killSwitch, packageEntriesStreamDone) = indexService
+        .packageEntries(None)
+        .viaMat(KillSwitches.single)(Keep.right[NotUsed, UniqueKillSwitch])
+        .toMat(Sink.foreach {
+          case PackageEntry.PackageUploadAccepted(submissionId, _) =>
+            packageSubmissionsTrackerMap(submissionId)._2.complete(Success(()))
+            ()
+          case PackageEntry.PackageUploadRejected(submissionId, _, reason) =>
+            packageSubmissionsTrackerMap(submissionId)._2
+              .complete(
+                Failure(new RuntimeException(s"Package upload at initialization failed: $reason"))
               )
-              .asScala
-          } yield ()
+            ()
+        })(Keep.both[UniqueKillSwitch, Future[Done]])
+        .run()
+
+      val uploadAndWaitPackages = Future.traverse(packageSubmissionsTrackerMap.toVector) {
+        case (submissionId, (file, promise)) =>
+          Future
+            .fromTry(DarParser.readArchiveFromFile(file).toTry)
+            .flatMap(dar =>
+              writeService
+                .uploadPackages(Ref.SubmissionId.assertFromString(submissionId), dar.all, None)(
+                  LoggingContext.ForTesting,
+                  NoOpTelemetryContext,
+                )
+                .asScala
+            )
+            .flatMap(_ => promise.future)
+      }
+
+      uploadAndWaitPackages
+        .map(_ => ())
+        .andThen { case _ =>
+          killSwitch.shutdown()
+          packageEntriesStreamDone
```

Comment:
I would prefer splitting this big method into smaller methods so it reads easier

Reply:
Went for conciseness. Fine by me though. Split